File: 2025/SoyoungJung/week2/2025-06-15-longest-repeating-character-replacement.ts
function characterReplacement(s: string, k: number): number {
  const freq: number[] = new Array(26).fill(0);
  let left = 0;
  let maxCount = 0;
  let maxLength = 0;

  for (let right = 0; right < s.length; right++) {
    const idx = s.charCodeAt(right) - 65;
    freq[idx]++;
    maxCount = Math.max(maxCount, freq[idx]);

    const currentWindowSize = right - left + 1;

    if (currentWindowSize - maxCount > k) {
      const leftIdx = s.charCodeAt(left) - 65;
      freq[leftIdx]--;
      left++;
    }

    maxLength = Math.max(maxLength, right - left + 1);
  }

  return maxLength;
}

File: 2025/jimmycho/week2/2025-06-15-Longest-Repeating-Character-Replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        # Declare variables to store character counts, maximum length of substring, and sliding window indices
        char_count = {s[0] : 1}
        max_length = 0
        left, right = 0, 1
        # Iterate process of incrementing window indices while upperbound is less than the length of string
        while right < len(s):
            # Calculate the max frequency of characters
            most_frequent = max(char_count.values())
            # Check if k is bigger than the number of characters other than the most frequent
            if right - left - most_frequent <= k:
                # Update max length
                max_length = max(max_length, right - left)
                # Increment upperbound and increment the character count
                right += 1
                char_count[s[right - 1]] = char_count.get(s[right - 1], 0) + 1
            else:
                # Decrement lowerbound character count and index
                char_count[s[left]] -= 1
                left += 1
        # One additional calculation outside loop
        most_frequent = max(char_count.values())
        if right - left - most_frequent <= k:
            max_length = max(max_length, right - left)
        # Return max length of substring
        return max_length
